derive/src/js_enum.rs

```diff
@@ -0,0 +1,255 @@
+extern crate proc_macro;
+
+use {proc_macro::TokenStream, quote::quote};
+
+pub fn js_enum(input: TokenStream) -> TokenStream {
+    let ast: syn::DeriveInput = syn::parse(input).unwrap();
+
+    if let syn::Data::Enum(enum_data) = &ast.data {
+        struct Variant {
+            name: String,
+            in_fields: Vec<syn::Ident>,
+            out_fields: Vec<(syn::Ident, syn::Type)>,
+        }
+
+        let mut variants = Vec::new();
+        let mut is_unnamed = false;
+
+        let ident = &ast.ident;
+        for variant in &enum_data.variants {
```

Comment:
this can be inlined with a `map` and `collect`, also consider extracting long expressions into nameable functions